File: scripts/serve.py
#!/usr/bin/env python3
"""Serve wiki engine assets overlaid with a live content directory."""

from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path
from urllib.parse import unquote, urlsplit
import mimetypes
import sys

ENGINE_ROOT = Path(__file__).resolve().parent.parent
CONTENT_ROOT = Path(sys.argv[1] if len(sys.argv) > 1 else ".").resolve()
PORT = int(sys.argv[2] if len(sys.argv) > 2 else "8080")

for required in ("_config.md", "_sidebar.md"):
    if not (CONTENT_ROOT / required).is_file():
        raise SystemExit(f"Wiki content is missing {required}: {CONTENT_ROOT}")


def safe_path(root, relative):
    candidate = (root / relative.lstrip("/")).resolve()
    try:
        candidate.relative_to(root)
        return candidate
    except ValueError:
        return None


def requested_file(pathname):
    if pathname == "/content" or pathname.startswith("/content/"):
        return safe_path(CONTENT_ROOT, pathname[len("/content"):])
    if pathname in ("/custom.css", "/favicon.svg"):
        override = CONTENT_ROOT / pathname[1:]
        if override.is_file():
            return override
    return safe_path(ENGINE_ROOT, pathname)


class WikiHandler(BaseHTTPRequestHandler):
    def do_HEAD(self):
        self.serve(send_body=False)

    def do_GET(self):
        self.serve(send_body=True)

    def serve(self, send_body):
        try:
            pathname = unquote(urlsplit(self.path).path)
            file = requested_file(pathname)
            if file is None:
                self.send_error(403)
                return
            if file.is_dir():
                file = file / "index.html"
            if not file.is_file():
                self.send_error(404)
                return

            size = file.stat().st_size
            content_type = mimetypes.guess_type(str(file))[0] or "application/octet-stream"
            if file.suffix == ".po":
                content_type = "text/x-gettext-translation"
            if content_type.startswith("text/") or file.suffix in (".js", ".json", ".md", ".svg"):
                content_type += "; charset=utf-8"
            self.send_response(200)
            self.send_header("Cache-Control", "no-store")
            self.send_header("Content-Length", str(size))
            self.send_header("Content-Type", content_type)
            self.end_headers()
            if send_body:
                with file.open("rb") as source:
                    self.wfile.write(source.read())
        except (OSError, ValueError):
            self.send_error(404)


print(f"Wiki available at http://localhost:{PORT}")
print(f"Serving live content from {CONTENT_ROOT}")
ThreadingHTTPServer(("127.0.0.1", PORT), WikiHandler).serve_forever()

File: scripts/serve.mjs
import { createReadStream, existsSync } from "node:fs"
import { stat } from "node:fs/promises"
import { createServer } from "node:http"
import { dirname, extname, resolve, sep } from "node:path"
import { fileURLToPath } from "node:url"

const engineRoot = resolve(dirname(fileURLToPath(import.meta.url)), "..")
const contentRoot = resolve(process.argv[2] || ".")
const port = Number(process.argv[3] || 8080)
const mimeTypes = {
    ".css": "text/css; charset=utf-8",
    ".gif": "image/gif",
    ".html": "text/html; charset=utf-8",
    ".jpeg": "image/jpeg",
    ".jpg": "image/jpeg",
    ".js": "text/javascript; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".md": "text/markdown; charset=utf-8",
    ".png": "image/png",
    ".svg": "image/svg+xml; charset=utf-8",
    ".ttf": "font/ttf",
    ".webp": "image/webp",
    ".woff": "font/woff",
    ".woff2": "font/woff2",
}

for (const required of ["_config.md", "_sidebar.md"]) {
    if (!existsSync(resolve(contentRoot, required))) {
        console.error(`Wiki content is missing ${required}: ${contentRoot}`)
        process.exit(1)
    }
}

function safePath(root, relativePath) {
    const file = resolve(root, relativePath)
    return file === root || file.startsWith(`${root}${sep}`) ? file : null
}

function requestedFile(pathname) {
    if (pathname === "/content" || pathname.startsWith("/content/")) {
        return safePath(contentRoot, pathname.slice("/content".length).replace(/^\/+/, ""))
    }
    if (pathname === "/custom.css" && existsSync(resolve(contentRoot, "custom.css"))) {
        return resolve(contentRoot, "custom.css")
    }
    if (pathname === "/favicon.svg" && existsSync(resolve(contentRoot, "favicon.svg"))) {
        return resolve(contentRoot, "favicon.svg")
    }
    return safePath(engineRoot, `.${pathname}`)
}

createServer(async (request, response) => {
    try {
        const pathname = decodeURIComponent(new URL(request.url, "http://localhost").pathname)
        let file = requestedFile(pathname)
        if (!file) {
            response.writeHead(403).end("Forbidden\n")
            return
        }

        let info = await stat(file)
        if (info.isDirectory()) {
            file = resolve(file, "index.html")
            info = await stat(file)
        }
        if (!info.isFile()) throw new Error("Not a file")

        response.writeHead(200, {
            "Cache-Control": "no-store",
            "Content-Length": info.size,
            "Content-Type": mimeTypes[extname(file).toLowerCase()] || "application/octet-stream",
        })
        if (request.method === "HEAD") response.end()
        else createReadStream(file).pipe(response)
    } catch {
        response.writeHead(404, {
            "Cache-Control": "no-store",
            "Content-Type": "text/plain; charset=utf-8",
        }).end("Not found\n")
    }
}).listen(port, "127.0.0.1", () => {
    console.log(`Wiki available at http://localhost:${port}`)
    console.log(`Serving live content from ${contentRoot}`)
})
